handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,526 @@
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ServerOcspContext serverOcspContext = newServerOcspContext();
+        TestClientOcspContext clientOcspContext = new TestClientOcspContext(true);
+
+        handshake(sslProvider, latch, serverOcspContext, clientOcspContext, serverHandler, clientHandler);
+
+        byte[] expected = serverOcspContext.response();
+        byte[] actual = clientOcspContext.response();
+
+        assertNotNull(actual);
+        assertNotSame(expected, actual);
+        assertArrayEquals(expected, actual);
+    }
+
+    @Test(timeout = 10000L)
+    public void testClientRejectingOcspStapleOpenSsl() throws Exception {
+        testClientRejectingOcspStaple(SslProvider.OPENSSL);
+    }
+
+    @Test(timeout = 10000L)
+    public void testClientRejectingOcspStapleOpenSslRefCnt() throws Exception {
+        testClientRejectingOcspStaple(SslProvider.OPENSSL_REFCNT);
+    }
+
+    /**
+     * The Server provides an OCSP staple and the Client rejects it.
+     */
+    private void testClientRejectingOcspStaple(SslProvider sslProvider) throws Exception {
+        final AtomicReference<Throwable> clientCauseRef = new AtomicReference<Throwable>();
+        final AtomicReference<Throwable> serverCauseRef = new AtomicReference<Throwable>();
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        ChannelInboundHandlerAdapter serverHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    serverCauseRef.set(cause);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    clientCauseRef.set(cause);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ServerOcspContext serverOcspContext = newServerOcspContext();
+        TestClientOcspContext clientOcspContext = new TestClientOcspContext(false);
+
+        handshake(sslProvider, latch, serverOcspContext, clientOcspContext, serverHandler, clientHandler);
+
+        byte[] expected = serverOcspContext.response();
+        byte[] actual = clientOcspContext.response();
+
+        assertNotNull(actual);
+        assertNotSame(expected, actual);
+        assertArrayEquals(expected, actual);
+
+        // If the client rejects the staple then Client and Server will fail with SSLHandshakeExceptions
+        assertTrue("" + clientCauseRef.get(), isCausedBy(clientCauseRef.get(), SSLHandshakeException.class));
```

Comment:
consider saving a reference to this exception ... just to ensure the same value is used in the comparison and string printout.

(general comment for this test)